fido/Crypto/Fido2/PublicKey.hs

```diff
@@ -0,0 +1,279 @@
+-- https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-client-to-authenticator-protocol-v2.0-id-20180227.html#ctap2-canonical-cbor-encoding-form
+decodePublicKey :: Decoder s PublicKey
+decodePublicKey = do
+  _n <- CBOR.decodeMapLenCanonical
+  ktyKey <- CBOR.decodeIntCanonical
+  when (ktyKey /= 1) $ fail "Expected required `kty`"
+  kty <- CBOR.decodeIntCanonical
+  case kty of
+    1 -> EdDSAPublicKey <$> decodeEdDSAKey
+    2 -> ECDSAPublicKey <$> decodeECDSAPublicKey
+    x -> fail $ "unexpected kty: " ++ show x
+
+decodeEdDSAKey :: Decoder s EdDSAKey
+decodeEdDSAKey = do
+  algKey <- CBOR.decodeIntCanonical
+  when (algKey /= 3) $ fail "Expected required `alg`"
+  alg <- decodeCOSEAlgorithmIdentifier
+  when (alg /= EdDSA) $ fail "Unsupported `alg`"
+  crvKey <- CBOR.decodeIntCanonical
+  when (crvKey /= (-1)) $ fail "Expected required `crv`"
+  crv <- CBOR.decodeIntCanonical
+  xKey <- CBOR.decodeIntCanonical
+  when (xKey /= -2) $ fail "Expected required `x`"
+  x <- CBOR.decodeBytesCanonical
+  case crv of
+    6 ->
+      case Ed25519.publicKey x of
+        CryptoFailed e -> fail (show e)
+        CryptoPassed a -> pure $ Ed25519 a
+    7 ->
+      case Ed448.publicKey x of
+        CryptoFailed e -> fail (show e)
+        CryptoPassed a -> pure $ Ed448 a
+    _ -> fail "Unsupported `crv`"
+
+decodeECDSAPublicKey :: Decoder s ECDSAKey
+decodeECDSAPublicKey = do
+  algKey <- CBOR.decodeIntCanonical
+  when (algKey /= 3) $ fail "Expected required `alg`"
+  alg <- decodeCOSEAlgorithmIdentifier
+  hash <- case alg of
+    ECDSAIdentifier x -> pure x
+    _ -> fail "Unsupported `alg`"
+  crvKey <- CBOR.decodeIntCanonical
+  when (crvKey /= (-1)) $ fail "Expected required `crv`"
+  crv <- CBOR.decodeIntCanonical
+  curve <- case crv of
+    1 -> pure $ ECC.getCurveByName ECC.SEC_p256r1
+    2 -> pure $ ECC.getCurveByName ECC.SEC_p384r1
+    3 -> pure $ ECC.getCurveByName ECC.SEC_p521r1
+    _ -> fail "Unsupported `crv`"
+  xKey <- CBOR.decodeIntCanonical
+  when (xKey /= -2) $ fail "Expected required `x`"
+  x <- os2ip <$> CBOR.decodeBytesCanonical
+  yKey <- CBOR.decodeIntCanonical
+  when (yKey /= -3) $ fail "Expected required `x`"
+  tokenType <- CBOR.peekTokenType
+  y <- case tokenType of
+    -- TODO(arianvp): Implement compressed curve. Waiting for
+    -- https://github.com/haskell-crypto/cryptonite/issues/302
+    CBOR.TypeBool -> fail "Compressed format not supported _yet_ See Issue number X"
+    -- direct coordinate
+    CBOR.TypeBytes -> os2ip <$> CBOR.decodeBytesCanonical
+    _ -> fail "Unexpected token type"
+  let point = ECC.Point x y
+  when (not (ECC.isPointValid curve point)) $ fail "point not on curve"
+  pure $ ECDSAKey hash (ECDSA.PublicKey curve (ECC.Point x y))
+
+encodePublicKey :: PublicKey -> Encoding
+encodePublicKey (ECDSAPublicKey (ECDSAKey alg (ECDSA.PublicKey curve point))) =
+  CBOR.encodeMapLen 5
+    <> CBOR.encodeInt 1
+    <> CBOR.encodeInt 2
+    <> CBOR.encodeInt 3
```

Comment:
There are a lot of magic values here. It would be helpful, both for verification, and for maintainability in case new key types need to be supported, to give them a name in a let-binding.